### PR TITLE
Draft agency API spec

### PR DESCRIPTION
**Pull request by adborden**
Resolves https://github.com/18F/foia/issues/32

Thanks @bsweger for drafting the bulk of this and consulting on the final touches. Once merged, this is ready for a first round of feedback from agencies.

### CODE REVIEW

**Comment by adborden** [comment]
I added a section for the 10% fields: Agency form fields. I think it makes sense to capture these in a separate field for namespacing. It also let's us talk about how they are different than other fields because they are defined in the metadata file.

File: agency-api-spec.md [hunk 1]
```diff
@@ -157,6 +157,118 @@ JSON payload that contains the form fields.
...
+
+Your agency might have additional fields specified in your [agency metadata file](https://github.com/18F/foia-recommendations/blob/master/schemas.md#agency-metadata-file).
+These additional fields are unique to your agency and are captured separately in
+the `agency_form_fields` object.
+
+The specific fields in `agency_form_fields` will be defined by your metadata
+file which includes both required and optional form fields. Any fields in
+`required_form_fields` will be considered required. Any fields in
+`additional_form_fields` will be considered optional. The portal will validate
+that any required fields are present.
```
Comment (bsweger) [inline agency-api-spec.md]: Do you think it's worth clarifying what this means from the agency perspective and/or being clearer that this validation isn't an implementation detail related to the process exposing this API?

Maybe rewording this last sentence to something like _DOJ's central FOIA portal will ensure that required fields are present before POSTing a request to your endpoint._

File: agency-api-spec.md [hunk 2]
```diff
@@ -157,6 +157,118 @@ JSON payload that contains the form fields.
...
+    // ...
+    "agency_form_fields": {
+        "contract_number": "5547",
+        "region": "9",
+        "request_origin": "Individual/Self"
+    }
+}
+```
+
+**Field:** | `request_origin`
```
Comment (bsweger) [inline agency-api-spec.md]: My initial reactive to having these agency-specific fields listed here in addition to the examples above is that this more formal presentation of them might give the impression that they're required universally. 

I think your metadata to JSON example above sufficiently conveys how these agency-specific questions will be handled, so I'd be inclined to remove the tables below (fair warning: when on the fence, I always default to less instead of more).

File: agency-api-spec.md [hunk 3]
```diff
@@ -221,3 +333,6 @@ Services like [api.data.gov](https://api.data.gov/about/) provide this authentic
 	}
```
Comment (bsweger) [inline agency-api-spec.md]: The sample call is looking good--thanks for taking the time to add a realistic payload. Do you think the agency-specific fields should also be a part of this sample?

**Review by bsweger — APPROVED** [review]
Thanks for the updates, @adborden. You're right to address the "10%" in this spec, and I really like the GSA example you added.

Left a few minor questions below, but my one larger piece of feedback is...I can't shake this feeling that we're making an artificial distinction between the types of FOIA questions/answers being passed to the API.
 We talk internally a lot about the 90% and the 10%,
 but is that meaningful to the recipient of the request?

One idea, if you want to maintain this distinction, might be to consider a different name for `agency_form_fields`. That's throwing me off a bit...for example, `agency` and `agency_component_name` are not `agency_form_fields`, but `contract_number` is. 

Maybe something like `component_specific_fields` or `agency_specific_fields`? To make it clearer that these aren't form fields and they aren't about the agency; rather it's info needed to process a FOIA request for this particular agency component.

Am marking this as approved because even if we don't resolve these questions, this is in good shape to shop around and get feedback on.

File: agency-api-spec.md (hunk 3 above)
Comment (adborden) [inline agency-api-spec.md]: Good call, I'll add it.

File: agency-api-spec.md (hunk 2 above)
Comment (adborden) [inline agency-api-spec.md]: Yeah, that's fair. I'll remove them.

**Review by hbillings — APPROVED** [review]
This looks good to me! Clear, helpful examples and language.

**Comment by adborden** [comment]
> I can't shake this feeling that we're making an artificial distinction between the types of FOIA questions/answers being passed to the API. We talk internally a lot about the 90% and the 10%, but is that meaningful to the recipient of the request?

I agree, it is a somewhat artificial distinction. I think maybe the distinction is important for technical reasons, but it would be good to use better language in that case too.

@RyanSibley not sure if you've had any thoughts on better names for the "90% fields" and "10% fields". For now, I'm calling the 10% "agency component specific fields" and I'm not giving any name to the 90%, they are just request fields.

**Comment by adborden** [comment]
@ba66e77 @malikkotob I've merged this but it is still a draft. Please comment more on this PR if you have any feedback. I'll be out next week but happy to address it when I return.

**Comment by malikkotob** [comment]
Thanks for your hard work here, y'all!

@adborden here are the actions for us to take as a result of our discussion on Thursday of last week:
1. Modify the headline "Make FOIA Request", since this API is more so about how agency components will handle FOIA requests that have been made, but not how to make them.
1. Under the URL Params section, it states that `id` is required. We should find a way to rephrase to denote that this is a 'suggested' format, or something similar, since our system won't actually enforce this approach.
1. Flatten the "Request fields" and "Agency component specific form fields" section into a single "Request fields" section.
1. Modify the agency metadata file example to match the format that JSON API generates for us (which includes not differentiating between required/non-required fields).

**Review by robertomiramon50-stack — APPROVED** [review]
